src/ripple/app/tx/impl/Escrow.cpp

```diff
@@ -198,17 +210,70 @@ EscrowCreate::doApply()
     if (!sle)
         return tefINTERNAL;
 
-    // Check reserve and funds availability
-    {
-        auto const balance = STAmount((*sle)[sfBalance]).xrp();
-        auto const reserve =
-            ctx_.view().fees().accountReserve((*sle)[sfOwnerCount] + 1);
+    STAmount const amount{ctx_.tx[sfAmount]};
+
+    std::shared_ptr<SLE> sleLine;
+
+    auto const balance = STAmount((*sle)[sfBalance]).xrp();
+    auto const reserve =
+        ctx_.view().fees().accountReserve((*sle)[sfOwnerCount] + 1);
+    bool isIssuer = amount.getIssuer() == account;
 
-        if (balance < reserve)
-            return tecINSUFFICIENT_RESERVE;
+    if (balance < reserve)
+        return tecINSUFFICIENT_RESERVE;
 
+    // Check reserve and funds availability
+    if (isXRP(amount))
+    {
         if (balance < reserve + STAmount(ctx_.tx[sfAmount]).xrp())
             return tecUNFUNDED;
+        // pass
+    }
+    else
+    {
+        // preflight will prevent this ever firing, included
+        // defensively for completeness
+        if (!ctx_.view().rules().enabled(featurePaychanAndEscrowForTokens))
+            return temDISABLED;
+
+        TER result = trustTransferAllowed(
+            ctx_.view(),
+            {account, ctx_.tx[sfDestination]},
+            amount.issue(),
+            ctx_.journal);
+
+        JLOG(ctx_.journal.trace())
+            << "EscrowCreate::doApply trustTransferAllowed result=" << result;
+
+        if (!isTesSuccess(result))
+            return result;
+
+        // issuer does not need to lock anything
+        if (!isIssuer)
+        {
+            // perform the lock as a dry run before
+            // we modify anything on-ledger
+            sleLine = ctx_.view().peek(keylet::line(
+                account, amount.getIssuer(), amount.getCurrency()));
+
+            // check if the escrow is capable of being
+            // finished before we allow it to be created
+            if (!sleLine)
```

Comment:
wouldn't this make sense to create a `preclaim` for `EscrowCreate`? (i'm not too familiar with escrow design not take it with grain of salt)

Reply:
I think leaving it as it is fine 👍